Please update vue-demi dependency to 0.13.x for Vue 2.7 support

Vue 2.7 have composition API integrated in the package, but vuelidate require @vue/composition-api, now i need to keep both for properly work and have some side effects from that.

The following error occurs when adding vuelidate to a fresh Vue 2.7 project.

```
Compiled with problems:X

ERROR in ./node_modules/vue-demi/lib/index.mjs 2:0-81

Module not found: Error: Can't resolve '@vue/composition-api/dist/vue-composition-api.mjs' in 'D:\Development\Vue\vue-app27\node_modules\vue-demi\lib'
```

Package.json snippet:
+-- vue@2.7.2
+-- @vuelidate/core@2.0.0-alpha.42
+-- @vuelidate/validators@2.0.0-alpha.30

> The following error occurs when adding vuelidate to a fresh Vue 2.7 project.
> 
> ```
> Compiled with problems:X
> 
> ERROR in ./node_modules/vue-demi/lib/index.mjs 2:0-81
> 
> Module not found: Error: Can't resolve '@vue/composition-api/dist/vue-composition-api.mjs' in 'D:\Development\Vue\vue-app27\node_modules\vue-demi\lib'
> ```
> 
> Package.json snippet: +-- vue@2.7.2 +-- @vuelidate/core@2.0.0-alpha.42 +-- @vuelidate/validators@2.0.0-alpha.30

You can fix this temporarily by overriding dependency with this rule in your package.json file, but again, you need to have installed @vue/composition-api because Vuelidate currently can't use composition-api from Vue 2.7 itself.

`  "overrides": {
    "vue-demi": "npm:vue-demi@^0.13.1"
  }`

--- a/package.json
+++ b/package.json
@@ -46,7 +46,8 @@
     "standard-version": "^9.3.1",
     "vue": "^3.2.2",
     "vue-template-compiler": "^2.6.14",
-    "vue2": "npm:vue@2.6.14"
+    "vue2": "npm:vue@2.6.14",
+    "vue2.7": "npm:vue@2.7.5"
   },
   "scripts": {
     "lerna:publish": "lerna publish --conventional-commits",
@@ -60,9 +61,11 @@
     "release:dry": "yarn lerna:build && yarn lerna:version --no-push --no-git-tag-version",
     "test:unit": "npm run test:3",
     "test:2": "npm run use-vue:2 && jest",
+    "test:2.7": "npm run use-vue:2.7 && jest",
     "test:3": "npm run use-vue:3 && jest",
-    "test:all": "npm run test:2 && npm run test:3",
+    "test:all": "npm run test:2 && npm run test:2.7 && npm run test:3",
     "use-vue:2": "node scripts/swap-vue.js 2 && vue-demi-switch 2",
+    "use-vue:2.7": "node scripts/swap-vue.js 2.7 && vue-demi-switch 2.7",
     "use-vue:3": "node scripts/swap-vue.js 3 && vue-demi-switch 3"
   },
   "config": {

--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     "url": "https://github.com/vuelidate/vuelidate.git"
   },
   "dependencies": {
-    "vue-demi": "^0.12.0"
+    "vue-demi": "^0.13.4"
   },
   "optionalDependencies": {
     "@vue/composition-api": "^1.3.1"

--- a/package.json
+++ b/package.json
@@ -46,6 +46,7 @@
     "standard-version": "^9.3.1",
     "vue": "^3.2.2",
     "vue-template-compiler": "^2.6.14",
+    "vue-template-compiler2.7": "npm:vue-template-compiler@2.7.5",
     "vue2": "npm:vue@2.6.14",
     "vue2.7": "npm:vue@2.7.5"
   },

--- a/packages/docs/src/advanced_usage.md
+++ b/packages/docs/src/advanced_usage.md
@@ -779,7 +779,7 @@ the `currentVueInstance` config to pass the component's vue instance.
 export default {
   render: () => {},
   async setup () {
-    const currentVueInstance = getCurrentInstance()
+    const currentVueInstance = getCurrentInstance()?.proxy
     const result = await doAsyncStuff()
     const vuelidate = useVuelidate(rules, result.state, { currentVueInstance })
     return { vuelidate }

--- a/packages/validators/package.json
+++ b/packages/validators/package.json
@@ -20,7 +20,7 @@
     "lint": "eslint src"
   },
   "dependencies": {
-    "vue-demi": "^0.12.0"
+    "vue-demi": "^0.13.4"
   },
   "devDependencies": {
     "bili": "^5.0.5",

--- a/packages/vuelidate/package.json
+++ b/packages/vuelidate/package.json
@@ -31,7 +31,7 @@
     "dev": "yarn build --watch"
   },
   "dependencies": {
-    "vue-demi": "^0.12.0"
+    "vue-demi": "^0.13.4"
   },
   "devDependencies": {
     "@babel/core": "^7.15.0",

--- a/packages/vuelidate/src/index.js
+++ b/packages/vuelidate/src/index.js
@@ -34,10 +34,10 @@ export function useVuelidate (validations, state, globalConfig = {}) {
   }
   let { $registerAs, $scope = CollectFlag.COLLECT_ALL, $stopPropagation, $externalResults, currentVueInstance } = globalConfig
 
-  const instance = currentVueInstance || getCurrentInstance()
+  const instance = currentVueInstance || getCurrentInstance()?.proxy
 
   const componentOptions = instance
-    ? instance.proxy.$options
+    ? instance.$options
     : {}
   // if there is no registration name, add one.
   if (!$registerAs && instance) {
@@ -55,7 +55,7 @@ export function useVuelidate (validations, state, globalConfig = {}) {
     sendValidationResultsToParent,
     removeValidationResultsFromParent
   } = instance
-    ? nestedValidations({ $scope, instance: instance.proxy })
+    ? nestedValidations({ $scope, instance })
     : { childResults: ref({}) }
 
   // Options API
@@ -66,7 +66,7 @@ export function useVuelidate (validations, state, globalConfig = {}) {
     onBeforeMount(() => {
       // Delay binding state to validations defined with the Options API until mounting, when the data
       // has been attached to the component instance. From that point on it will be reactive.
-      state.value = instance.proxy
+      state.value = instance
 
       watch(() => isFunction(rules) ? rules.call(state.value, new ComputedProxyFactory(state.value)) : rules,
         (validations) => {
@@ -76,8 +76,8 @@ export function useVuelidate (validations, state, globalConfig = {}) {
             childResults,
             resultsCache,
             globalConfig,
-            instance: instance.proxy,
-            externalResults: $externalResults || instance.proxy.vuelidateExternalResults
+            instance,
+            externalResults: $externalResults || instance.vuelidateExternalResults
           })
         }, { immediate: true })
     })
@@ -96,7 +96,7 @@ export function useVuelidate (validations, state, globalConfig = {}) {
         childResults,
         resultsCache,
         globalConfig,
-        instance: instance ? instance.proxy : {},
+        instance: instance ?? {},
         externalResults: $externalResults
       })
     }, { immediate: true })

--- a/packages/vuelidate/src/index.js
+++ b/packages/vuelidate/src/index.js
@@ -55,7 +55,7 @@ export function useVuelidate (validations, state, globalConfig = {}) {
     sendValidationResultsToParent,
     removeValidationResultsFromParent
   } = instance
-    ? nestedValidations({ $scope, instance })
+    ? nestedValidations({ $scope, instance: instance.proxy })
     : { childResults: ref({}) }
 
   // Options API

--- a/scripts/swap-vue.js
+++ b/scripts/swap-vue.js
@@ -62,7 +62,6 @@ function resetPackageNames () {
 }
 
 function useTemplateCompilerVersion (version) {
-  console.log('existing!! ? ', !fs.existsSync(vueTemplateCompiler))
   if (!fs.existsSync(vueTemplateCompiler)) {
     console.log('There is no default vue-template-compiler version, finding it')
     if (version === 2.7 && fs.existsSync(vueTemplateCompiler2_7)) {

--- a/scripts/swap-vue.js
+++ b/scripts/swap-vue.js
@@ -1,15 +1,13 @@
+/* eslint-disable camelcase */
 const fs = require('fs')
 const path = require('path')
 
 const Vue2 = path.join(__dirname, '../node_modules/vue2')
-// eslint-disable-next-line camelcase
 const Vue2_7 = path.join(__dirname, '../node_modules/vue2.7')
 const DefaultVue = path.join(__dirname, '../node_modules/vue')
 const Vue3 = path.join(__dirname, '../node_modules/vue3')
 const vueTemplateCompiler = path.join(__dirname, '../node_modules/vue-template-compiler')
-// eslint-disable-next-line camelcase
 const vueTemplateCompiler2_6 = path.join(__dirname, '../node_modules/vue-template-compiler2.6')
-// eslint-disable-next-line camelcase
 const vueTemplateCompiler2_7 = path.join(__dirname, '../node_modules/vue-template-compiler2.7')
 
 const version = Number(process.argv[2]) || 3

--- a/scripts/swap-vue.js
+++ b/scripts/swap-vue.js
@@ -2,6 +2,8 @@ const fs = require('fs')
 const path = require('path')
 
 const Vue2 = path.join(__dirname, '../node_modules/vue2')
+// eslint-disable-next-line camelcase
+const Vue2_7 = path.join(__dirname, '../node_modules/vue2.7')
 const DefaultVue = path.join(__dirname, '../node_modules/vue')
 const Vue3 = path.join(__dirname, '../node_modules/vue3')
 
@@ -15,6 +17,9 @@ function useVueVersion (version) {
     if (version === 2 && fs.existsSync(Vue3)) {
       rename(Vue3, DefaultVue)
       console.log('Renamed "vue3" to "vue"')
+    } else if (version === 2.7 && fs.existsSync(Vue2_7)) {
+      rename(Vue2_7, DefaultVue)
+      console.log('Renamed "vue2.7" to "vue"')
     } else {
       rename(Vue2, DefaultVue)
       console.log('Renamed "vue2" to "vue"')
@@ -22,13 +27,28 @@ function useVueVersion (version) {
   }
 
   if (version === 3 && fs.existsSync(Vue3)) {
-    rename(DefaultVue, Vue2)
+    resetPackageNames()
     rename(Vue3, DefaultVue)
+  } else if (version === 2.7 && fs.existsSync(Vue2_7)) {
+    resetPackageNames()
+    rename(Vue2_7, DefaultVue)
   } else if (version === 2 && fs.existsSync(Vue2)) {
-    rename(DefaultVue, Vue3)
+    resetPackageNames()
     rename(Vue2, DefaultVue)
   } else {
     console.log(`Vue ${version} is already in use`)
+  }
+}
+
+function resetPackageNames () {
+  if (!fs.existsSync(Vue3)) {
+    rename(DefaultVue, Vue3)
+  } else if (!fs.existsSync(Vue2_7)) {
+    rename(DefaultVue, Vue2_7)
+  } else if (!fs.existsSync(Vue2)) {
+    rename(DefaultVue, Vue2)
+  } else {
+    console.error('Unable to reset package names')
   }
 }
 

--- a/scripts/swap-vue.js
+++ b/scripts/swap-vue.js
@@ -83,11 +83,10 @@ function useTemplateCompilerVersion (version) {
 
 function rename (fromPath, toPath) {
   if (!fs.existsSync(fromPath)) return
-  fs.rename(fromPath, toPath, function (err) {
-    if (err) {
-      console.log(err)
-    } else {
-      console.log(`Successfully renamed ${fromPath} to ${toPath} .`)
-    }
-  })
+  try {
+    fs.renameSync(fromPath, toPath)
+    console.log(`Successfully renamed ${fromPath} to ${toPath} .`)
+  } catch (err) {
+    console.log(err)
+  }
 }

--- a/scripts/swap-vue.js
+++ b/scripts/swap-vue.js
@@ -6,6 +6,11 @@ const Vue2 = path.join(__dirname, '../node_modules/vue2')
 const Vue2_7 = path.join(__dirname, '../node_modules/vue2.7')
 const DefaultVue = path.join(__dirname, '../node_modules/vue')
 const Vue3 = path.join(__dirname, '../node_modules/vue3')
+const vueTemplateCompiler = path.join(__dirname, '../node_modules/vue-template-compiler')
+// eslint-disable-next-line camelcase
+const vueTemplateCompiler2_6 = path.join(__dirname, '../node_modules/vue-template-compiler2.6')
+// eslint-disable-next-line camelcase
+const vueTemplateCompiler2_7 = path.join(__dirname, '../node_modules/vue-template-compiler2.7')
 
 const version = Number(process.argv[2]) || 3
 
@@ -19,9 +24,11 @@ function useVueVersion (version) {
       console.log('Renamed "vue3" to "vue"')
     } else if (version === 2.7 && fs.existsSync(Vue2_7)) {
       rename(Vue2_7, DefaultVue)
+      useTemplateCompilerVersion(2.7)
       console.log('Renamed "vue2.7" to "vue"')
     } else {
       rename(Vue2, DefaultVue)
+      useTemplateCompilerVersion(2)
       console.log('Renamed "vue2" to "vue"')
     }
   }
@@ -32,9 +39,11 @@ function useVueVersion (version) {
   } else if (version === 2.7 && fs.existsSync(Vue2_7)) {
     resetPackageNames()
     rename(Vue2_7, DefaultVue)
+    useTemplateCompilerVersion(2.7)
   } else if (version === 2 && fs.existsSync(Vue2)) {
     resetPackageNames()
     rename(Vue2, DefaultVue)
+    useTemplateCompilerVersion(2)
   } else {
     console.log(`Vue ${version} is already in use`)
   }
@@ -52,13 +61,36 @@ function resetPackageNames () {
   }
 }
 
+function useTemplateCompilerVersion (version) {
+  console.log('existing!! ? ', !fs.existsSync(vueTemplateCompiler))
+  if (!fs.existsSync(vueTemplateCompiler)) {
+    console.log('There is no default vue-template-compiler version, finding it')
+    if (version === 2.7 && fs.existsSync(vueTemplateCompiler2_7)) {
+      rename(vueTemplateCompiler2_7, vueTemplateCompiler)
+      console.log('Renamed "vue-template-compliler2.7" to "vue-template-compliler"')
+    } else {
+      rename(vueTemplateCompiler2_6, vueTemplateCompiler)
+      console.log('Renamed "vue-template-compliler2.6" to "vue-template-compliler"')
+    }
+  }
+  if (version === 2.7 && fs.existsSync(vueTemplateCompiler2_7)) {
+    rename(vueTemplateCompiler, vueTemplateCompiler2_6)
+    rename(vueTemplateCompiler2_7, vueTemplateCompiler)
+  } else if (version === 2 && fs.existsSync(vueTemplateCompiler2_6)) {
+    rename(vueTemplateCompiler, vueTemplateCompiler2_7)
+    rename(vueTemplateCompiler2_6, vueTemplateCompiler)
+  } else {
+    console.log(`vue-template-compliler ${version} is already in use`)
+  }
+}
+
 function rename (fromPath, toPath) {
   if (!fs.existsSync(fromPath)) return
   fs.rename(fromPath, toPath, function (err) {
     if (err) {
       console.log(err)
     } else {
-      console.log(`Successfully renamed ${fromPath} to ${toPath}.`)
+      console.log(`Successfully renamed ${fromPath} to ${toPath} .`)
     }
   })
 }

